### ISSUE TITLE
Shell cell execution failure 

### ISSUE THREAD

**Issue by sanketd11** [att/rcloud#1770]
This issue can be reproduced on RCloud local instance only
Steps to replicate:
- In a notebook having shell cell, add new R cell 
- Execute 'q()' function. Session will reconnect due to this
- Click on 'Reconnect' option

After session gets connected try to execute the shell cell. The shell cell will not be executed and will display following error messages 

In output div
![shell1](https://cloud.githubusercontent.com/assets/9251570/12030519/e84ebdfa-ae25-11e5-8acb-851fe82d9e25.PNG)

In session div

![shell2](https://cloud.githubusercontent.com/assets/9251570/12030520/f46dcb94-ae25-11e5-93cb-27f0995dab2d.PNG)

We need to restart the Rserve in order to execute shell cells again.

**Comment by s-u** [comment]
My guess is that `q()` will actually delete the R temporary directory so it won't exist afterwards. This will only affect installations without user switching. Note that shell cells are just one of many cases that will break when the temporary directory is gone - a lot of R code and packages rely on the temporary directory.

**Comment by s-u** [comment]
When setting up a session we should check that the directory exists. R does that check on startup but we should do it again.

Note that this is only borderline a bug - the behavior after running `q()` is undefined. In a single-user setup you can equally simply blow away things from your session and kill anyone's session if you so desire so it's really not advisable to run things that way.

**Comment by sanketd11** [comment]
Some more features affected by this issue are -
- shiny.html

![shiny](https://cloud.githubusercontent.com/assets/9251570/12136593/6e0e43b2-b46f-11e5-93ac-dd7592742c11.PNG)

-help

![help](https://cloud.githubusercontent.com/assets/9251570/12136592/6e0df4e8-b46f-11e5-90b7-c8fa98e15187.PNG)

**Comment by s-u** [comment]
Yes, like said, this breaks a lot of things in R (nothing to do with RCloud) since a lot of R code relies on the existence of the temporary directory.

**Comment by gordonwoodhull** [comment]
Is there some extra initialization we can do for each session? Wouldn't folks usually want a fresh temp directory with each session?

**Comment by s-u** [comment]
Yes we can and we do when user switching is enabled so that each user has its own temp dir (for permission reasons, really). We don't do that in non-switched setup. Partially, because R doesn't actually support that (there is no official way to set the tempdir in an existing session, because R assumes it has it under control directly). Note that this is R-itnernal directory and R manages the files that - it's different form the session temporary directory (which is truly session-specific).

**Comment by s-u** [comment]
Just to clarify this is nothing to do with the system `TEMPDIR` - that's a whole different issue, will fix the title. (BTW I don't expect any comments there - it's one line of code and we have discussed it more than it's worth it ;))

**Comment by s-u** [comment]
Actually, there is really no good way to fix this. The problem is that even if we guarantee the existence of `tempdir()` at session init, it doesn't help as another session can blow it away at any other point. So I'm inclined to simply override `q()` and document "don't do it" since there is no way for us to prevent the user from doing stupid things like that.

**Comment by gordonwoodhull** [comment]
My point was why should the sessions share the tempdir when they are separate processes. Seems like it's bound to run into other kinds of collisions. So why not always create a tempdir on session start? Or am I missing something?

**Comment by s-u** [comment]
Because R is managing that at R start but the server is forking so at the time we get into session code it's already established by R that was started as the server. If we do something ourselves then we have to maintain it (cleanup etc.). If we think this is a problem we'll probably have to do something along the lines of creating a new tree, overriding the R internals to replace the R managed one and then add cleanup. And that won't help if someone just deletes it anyway... so I'm not sure I really want to go deep down that road.

**Comment by gordonwoodhull** [comment]
@s-u, didn't you address part of this?
